objstore: Added WithExpectedErrs which allows to control instrumentation (e.g not increment failures for expected not found).

bwplotka opened the pull request:
This 3rd alternative to below PRs that was discussed offline: https://cloud-native.slack.com/archives/CL25937SP/p1585903230146500

Closes https://github.com/thanos-io/thanos/pull/2365 and https://github.com/thanos-io/thanos/pull/2369

This allows to not wake up on-call in the middle of the night, because of the expected, properly handled case (:

Thanks @pracucci and @pstibrany for starting this work and discussion. We were aware of this for like 12 months, but only with your discussions, we clarified what is expected here :hugs:  Your input was super valuable.

## Changes

Added new interface:

```go
type InstrumentedBucket interface {
	Bucket

	// WithExpectedErrs allows to specify a filter that marks certain errors as expected, so it will not wake up
	// on-call engineer (:
	WithExpectedErrs(IsOpFailureExpectedFunc) Bucket
```

Also: Has to move inmem to objstore for testing.

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

GiedriusS commented on pkg/objstore/azure/azure.go:
```diff
@@ -64,7 +64,7 @@ func (conf *Config) validate() error {
 }
 
-// NewBucket returns a new Bucket using the provided Azure config.
+// NewInMemBucket returns a new Bucket using the provided Azure config.
```
Seems like an unrelated change?

bwplotka replied:
:man_facepalming:

GiedriusS commented on pkg/objstore/client/factory.go:
```diff
@@ -40,9 +40,9 @@ type BucketConfig struct {
 }
 
-// NewBucket initializes and returns new object storage clients.
+// NewInMemBucket initializes and returns new object storage clients.
```
Seems like an unrelated change?

GiedriusS commented on pkg/objstore/filesystem/filesystem.go:
```diff
@@ -43,7 +43,7 @@ func NewBucketFromConfig(conf []byte) (*Bucket, error) {
 }
 
-// NewBucket returns a new filesystem.Bucket.
+// NewInMemBucket returns a new filesystem.Bucket.
```
Ditto.

GiedriusS commented on pkg/objstore/gcs/gcs.go:
```diff
@@ -41,7 +41,7 @@ type Bucket struct {
 }
 
-// NewBucket returns a new Bucket against the given bucket handle.
+// NewInMemBucket returns a new Bucket against the given bucket handle.
```
Ditto.

GiedriusS commented on pkg/objstore/oss/oss.go:
```diff
@@ -150,7 +150,7 @@ func (b *Bucket) ObjectSize(ctx context.Context, name string) (uint64, error) {
 }
 
-// NewBucket returns a new Bucket using the provided oss config values.
+// NewInMemBucket returns a new Bucket using the provided oss config values.
```
Ditto.

pracucci commented on pkg/block/fetcher.go:
```diff
@@ -236,7 +236,7 @@ func (f *BaseFetcher) loadMeta(ctx context.Context, id ulid.ULID) (*metadata.Met
 	}
 
-	r, err := f.bkt.Get(ctx, metaFile)
+	r, err := f.bkt.ReaderWithExpectedErrs(f.bkt.IsObjNotFoundErr).Get(ctx, metaFile)
```
Is this expected to cover the case of a partially uploaded block (while still uploading)?

bwplotka replied:
Yes

bwplotka commented:
Exactly this + Giedrius addressed comments was merged in release 0.12 and will be ported to master soon: https://github.com/thanos-io/thanos/commit/c05daeeace272a94fae81fd0c15ff70adab50e10
